compiler/rustc_trait_selection/src/solve/fulfill.rs

```diff
@@ -288,7 +293,10 @@ fn fulfillment_error_for_stalled<'tcx>(
     root_obligation: PredicateObligation<'tcx>,
 ) -> FulfillmentError<'tcx> {
     let (code, refine_obligation) = infcx.probe(|_| {
-        match infcx.evaluate_root_goal(root_obligation.clone().into(), GenerateProofTree::No).0 {
+        match <&SolverDelegate<'tcx>>::from(infcx)
+            .evaluate_root_goal(root_obligation.clone().into(), GenerateProofTree::No)
```

Comment:
This seems very meh, I would prefer to implement `evaluate_root_goal` and `evaluate_root_goal_raw` via an extension trait on `InferCtxt`. Why do we need the `SolverDelegate` super trait on that one?

Reply:
because `EvalCtxt::enter_root` requires a `SolverDelegate`